### DISCUSSION TITLE
hpx for web applications

### DISCUSSION THREAD

**Discussion by felixguendling** [STEllAR-GROUP/hpx#6483] [Q&A] [answered]
Hi everyone! 👋 😄 

I am the maintainer of [MOTIS](https://github.com/motis-project/motis) which is an open source mobility platform. The focus is door-to-door routing. Our core is based on the [RAPTOR algorithm](https://www.microsoft.com/en-us/research/wp-content/uploads/2012/01/raptor_alenex.pdf) which can be parallelised and we're planning to do this. We're currently using our own custom task library [ctx](https://github.com/motis-project/ctx) (with Boost context under the hood) which works fine for now but might not be the best option for parallelizing the routing algorithm.

Since MOTIS is a web application, it's important to not use all system resources to answer one query. Otherwise, one request that runs a bit longer will block all other user's requests. So my idea was to consider groups of 2 or 4 threads as one "processing unit" that processes a query in parallel. Ideally, those threads of one "processing unit" would be close to each other on the hardware (i.e. share one L2 cache). Then there's a global work stealing queue where these "processing units" get their work from. There needs to be an I/O thread to accept HTTP requests, put them into the queue. After finishing the request, the "processing unit" enqueues the response into the I/O thread so it can be sent back to the client.

Another topic is that in our custom ctx library we currently have one global work stealing queue with the option to append at the front of the queue (so it works like a stack, LIFO) or at the back (FIFO). I/O tasks such as new requests coming in via HTTP requests use FIFO but once a task runs, all subsequent tasks spawned by this task are enqueued LIFO so it finishes as fast as possible. Is there something similar in hpx?

I guess hpx is primarily used in batch-processing jobs on super computers where these problems do not exist. So my questions are: Does this approach make sense? Is hpx a good fit for this scenario? Can this be done with hpx and if so - how? Is there a best practice on how to connect hpx with a web server?

The documentation about [HPX runtime and resources](https://hpx-docs.stellar-group.org/latest/html/manual/hpx_runtime_and_resources.html) looks like it can be done. But maybe not everything at the same time? It looks like I have to chose between LIFO/FIFO for one queue and I can create a thread pool as my "processing unit" but then how do those thread pools steal work from a queue where the web server puts it?

**Comment by hkaiser** [comment, marked as answer]
All of the features you talked about are available in HPX. All of those can be controlled either through the command line or programmatically when initializing HPX:

- We do have lifo and fifo schedulers (currently with some platform-specific limitations, but we're working on resolving those)
- We can control the number of cores used for a scheduler (thread pool)
- You can create different thread pools in the same process (can be done programmatically only, no commend line support)
- HPX has dedicated (kernel-) threads that can be used for IO
- You can easily control where to launch tasks using executors (e.g. to run things either on the IO thread or on a particular thread pool)

I'd suggest you join our discord server to discuss things in detail (https://discord.gg/Tn9QuzVjvy).

**Reply by felixguendling** [reply]
Thank you very much for your quick response! Then I'll try to experiment a bit and ask concrete questions on Discord when they come up.
